File: cascadeflow/providers/together.py
"""Together.ai provider implementation."""

import os
import time
import math
from typing import Optional, Dict, Any

import httpx

from .base import BaseProvider, ModelResponse
from ..exceptions import ProviderError, ModelError


class TogetherProvider(BaseProvider):
    """
    Together.ai provider.

    Supports 50+ optimized open-source models.
    OpenAI-compatible API with full logprobs support.

    Example:
        >>> provider = TogetherProvider(api_key="...")
        >>> response = await provider.complete(
        ...     prompt="What is AI?",
        ...     model="meta-llama/Meta-Llama-3.1-8B-Instruct-Turbo"
        ... )

        >>> # With logprobs
        >>> response = await provider.complete(
        ...     prompt="What is AI?",
        ...     model="meta-llama/Meta-Llama-3.1-8B-Instruct-Turbo",
        ...     logprobs=True,
        ...     top_logprobs=10
        ... )
    """

    def __init__(self, api_key: Optional[str] = None):
        """
        Initialize Together.ai provider.

        Args:
            api_key: Together.ai API key. If None, reads from TOGETHER_API_KEY env var.
        """
        super().__init__(api_key)
        self.base_url = "https://api.together.xyz/v1"

        # Initialize HTTP client
        self.client = httpx.AsyncClient(
            headers={
                "Authorization": f"Bearer {self.api_key}",
                "Content-Type": "application/json"
            },
            timeout=60.0
        )

    def _load_api_key(self) -> Optional[str]:
        """Load API key from environment."""
        api_key = os.getenv("TOGETHER_API_KEY")
        if not api_key:
            raise ValueError(
                "Together.ai API key not found. Please set TOGETHER_API_KEY environment "
                "variable or pass api_key parameter."
            )
        return api_key

    def _check_logprobs_support(self) -> bool:
        """Together.ai supports logprobs (OpenAI-compatible)."""
        return True

    async def complete(
            self,
            prompt: str,
            model: str,
            max_tokens: int = 4096,
            temperature: float = 0.7,
            system_prompt: Optional[str] = None,
            **kwargs
    ) -> ModelResponse:
        """
        Complete a prompt using Together.ai API.

        Args:
            prompt: User prompt
            model: Model name (e.g., 'meta-llama/Meta-Llama-3.1-8B-Instruct-Turbo')
            max_tokens: Maximum tokens to generate
            temperature: Sampling temperature (0-1)
            system_prompt: Optional system prompt
            **kwargs: Additional Together.ai parameters
                     NEW: logprobs (bool) - Enable logprobs
                          top_logprobs (int) - Get top-k alternatives (1-20)

        Returns:
            ModelResponse with standardized format (enhanced with logprobs)

        Raises:
            ProviderError: If API call fails
            ModelError: If model execution fails
        """
        start_time = time.time()

        # Extract logprobs parameters
        logprobs_enabled = kwargs.pop('logprobs', False)
        top_logprobs = kwargs.pop('top_logprobs', None)

        # Build messages (OpenAI-compatible)
        messages = []
        if system_prompt:
            messages.append({"role": "system", "content": system_prompt})
        messages.append({"role": "user", "content": prompt})

        # Build request payload
        payload = {
            "model": model,
            "messages": messages,
            "max_tokens": max_tokens,
            "temperature": temperature,
            **kwargs
        }

        # Add logprobs if requested (OpenAI-compatible)
        if logprobs_enabled:
            payload["logprobs"] = True
            if top_logprobs:
                payload["top_logprobs"] = min(top_logprobs, 20)

        try:
            # Make API request
            response = await self.client.post(
                f"{self.base_url}/chat/completions",
                json=payload
            )
            response.raise_for_status()

            data = response.json()

            # Extract response (OpenAI-compatible format)
            choice = data["choices"][0]
            content = choice["message"]["content"]
            tokens_used = data["usage"]["total_tokens"]

            # Calculate latency
            latency_ms = (time.time() - start_time) * 1000

            # Calculate cost
            cost = self.estimate_cost(tokens_used, model)

            # Calculate confidence
            confidence = self.calculate_confidence(content, data)

            # Build base response
            model_response = ModelResponse(
                content=content,
                model=model,
                provider="together",
                cost=cost,
                tokens_used=tokens_used,
                confidence=confidence,
                latency_ms=latency_ms,
                metadata={
                    "finish_reason": choice["finish_reason"],
                    "prompt_tokens": data["usage"]["prompt_tokens"],
                    "completion_tokens": data["usage"]["completion_tokens"],
                }
            )

            # Parse logprobs if available (OpenAI-compatible format)
            if logprobs_enabled and "logprobs" in choice and choice["logprobs"]:
                logprobs_data = choice["logprobs"]

                if "content" in logprobs_data and logprobs_data["content"]:
                    tokens = []
                    logprobs_list = []
                    top_logprobs_list = []

                    for token_data in logprobs_data["content"]:
                        # Extract token
                        tokens.append(token_data["token"])

                        # Extract logprob
                        logprobs_list.append(token_data["logprob"])

                        # Extract top alternatives
                        if "top_logprobs" in token_data and token_data["top_logprobs"]:
                            top_k = {}
                            for alt in token_data["top_logprobs"]:
                                top_k[alt["token"]] = alt["logprob"]
                            top_logprobs_list.append(top_k)
                        else:
                            top_logprobs_list.append({})

                    # Add to response
                    model_response.tokens = tokens
                    model_response.logprobs = logprobs_list
                    model_response.top_logprobs = top_logprobs_list

                    # Update confidence based on actual probabilities
                    if logprobs_list:
                        avg_prob = sum(math.exp(lp) for lp in logprobs_list) / len(logprobs_list)
                        # Blend with heuristic confidence
                        model_response.confidence = (confidence + avg_prob) / 2

                    # Add metadata
                    model_response.metadata["has_logprobs"] = True
                    model_response.metadata["estimated"] = False
            else:
                # No logprobs available - add estimated values
                if logprobs_enabled:
                    model_response = self.add_logprobs_fallback(
                        model_response,
                        temperature,
                        base_confidence=0.82  # Together.ai models are good quality
                    )

            return model_response

        except httpx.HTTPStatusError as e:
            if e.response.status_code == 401:
                raise ProviderError(
                    "Invalid Together.ai API key",
                    provider="together",
                    original_error=e
                )
            elif e.response.status_code == 429:
                raise ProviderError(
                    "Together.ai rate limit exceeded",
                    provider="together",
                    original_error=e
                )
            else:
                raise ProviderError(
                    f"Together.ai API error: {e.response.status_code}",
                    provider="together",
                    original_error=e
                )
        except httpx.RequestError as e:
            raise ProviderError(
                "Failed to connect to Together.ai API",
                provider="together",
                original_error=e
            )
        except (KeyError, IndexError) as e:
            raise ModelError(
                f"Failed to parse Together.ai response: {e}",
                model=model,
                provider="together"
            )

    def estimate_cost(self, tokens: int, model: str) -> float:
        """
        Estimate cost for Together.ai models.

        Args:
            tokens: Total tokens
            model: Model name

        Returns:
            Estimated cost in USD
        """
        # Together.ai pricing (approximate, check their pricing page)
        # Varies by model size

        # Rough estimates:
        if "405B" in model or "70B" in model:
            # Large models: ~$0.0008/1K tokens
            return (tokens / 1000) * 0.0008
        elif "13B" in model or "8B" in model:
            # Medium models: ~$0.0002/1K tokens
            return (tokens / 1000) * 0.0002
        else:
            # Small models: ~$0.0001/1K tokens
            return (tokens / 1000) * 0.0001

    def calculate_confidence(
            self,
            response: str,
            metadata: Optional[Dict[str, Any]] = None
    ) -> float:
        """
        Calculate confidence score for Together.ai response.

        Args:
            response: Model response text
            metadata: Response metadata from API

        Returns:
            Confidence score (0-1)
        """
        # Start with base confidence
        confidence = super().calculate_confidence(response)

        if metadata:
            # Adjust based on finish_reason
            finish_reason = metadata.get("choices", [{}])[0].get("finish_reason")

            if finish_reason == "stop":
                confidence = min(1.0, confidence + 0.1)
            elif finish_reason == "length":
                confidence = max(0.5, confidence - 0.1)

        return confidence

    async def __aenter__(self):
        """Async context manager entry."""
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Async context manager exit."""
        await self.client.aclose()

File: packages/core/src/providers/together.ts
/**
 * Together AI provider implementation
 *
 * Supports: Llama 4, Llama 3.3, Llama 3.1, DeepSeek, Qwen, Mixtral, and many more
 * Pricing: Pay-per-token pricing (competitive rates)
 * Features: OpenAI-compatible API, full tool calling support
 *
 * Together AI offers 100+ optimized open-source models with fast inference
 */

import { BaseProvider, type ProviderRequest } from './base';
import type { ProviderResponse, Tool, Message } from '../types';
import type { ModelConfig } from '../config';

/**
 * Together AI pricing per 1M tokens (October 2025)
 * Source: https://www.together.ai/pricing
 * Format: [input_rate, output_rate] per 1M tokens
 */
const TOGETHER_PRICING: Record<string, [number, number]> = {
  // Llama 4 Series
  'llama-4-maverick': [0.27, 0.85],
  'llama-4-scout': [0.18, 0.59],

  // Llama 3.3 Series
  'llama-3.3-70b': [0.59, 0.79],
  'llama3.3-70b': [0.59, 0.79],

  // Llama 3.1 Series (most popular)
  'llama-3.1-405b': [3.5, 3.5],
  'llama-3.1-70b': [0.59, 0.79],
  'llama-3.1-8b': [0.05, 0.08],
  'llama3.1-405b': [3.5, 3.5],
  'llama3.1-70b': [0.59, 0.79],
  'llama3.1-8b': [0.05, 0.08],

  // Llama 3.2 Series
  'llama-3.2-90b': [1.2, 1.2],
  'llama-3.2-11b': [0.18, 0.18],
  'llama-3.2-8b': [0.1, 0.1],
  'llama-3.2-3b': [0.06, 0.06],
  'llama-3.2-1b': [0.06, 0.06],

  // Llama 3 Series
  'llama-3-70b': [0.59, 0.79],
  'llama-3-8b': [0.05, 0.08],
  'llama3-70b': [0.59, 0.79],
  'llama3-8b': [0.05, 0.08],

  // DeepSeek Series
  'deepseek-r1-distill-llama-70b': [2.0, 2.0],
  'deepseek-r1-distill-qwen-14b': [1.6, 1.6],
  'deepseek-r1-distill-qwen-1.5b': [0.18, 0.18],
  'deepseek-r1-throughput': [0.55, 2.19],
  'deepseek-r1': [3.0, 7.0],
  'deepseek-v3': [1.25, 1.25],

  // Qwen Series
  'qwen3-235b': [0.2, 0.6],
  'qwen3-coder-480b': [2.0, 2.0],
  'qwen2.5-72b': [1.2, 1.2],
  'qwen2.5-coder-32b': [0.8, 0.8],
  'qwen2.5-14b': [0.8, 0.8],
  'qwen2.5-7b': [0.3, 0.3],
  'qwq-32b': [1.2, 1.2],

  // Kimi Series
  'kimi-k2': [1.0, 3.0],

  // Mixtral Series
  'mixtral-8x7b': [0.6, 0.6],
  'mistral-saba': [0.79, 0.79],
};

/**
 * Together AI provider for fast inference of open-source models
 */
export class TogetherProvider extends BaseProvider {
  readonly name = 'together';
  private baseUrl = 'https://api.together.xyz/v1';

  constructor(config: ModelConfig) {
    super(config);
  }

  async generate(request: ProviderRequest): Promise<ProviderResponse> {
    try {
      const apiKey = this.getApiKey();
      const messages = this.normalizeMessages(request.messages);
      const togetherMessages = this.convertToTogetherMessages(messages, request.systemPrompt);
      const tools = request.tools ? this.convertTools(request.tools) : undefined;

      const response = await fetch(`${this.baseUrl}/chat/completions`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${apiKey}`,
        },
        body: JSON.stringify({
          model: request.model || this.config.name,
          messages: togetherMessages,
          max_tokens: request.maxTokens || this.config.maxTokens || 1000,
          temperature: request.temperature ?? this.config.temperature ?? 0.7,
          tools,
          ...request.extra,
        }),
      });

      if (!response.ok) {
        const error = await response.json().catch(() => ({ error: response.statusText }));
        throw new Error(`Together AI API error (${response.status}): ${JSON.stringify(error)}`);
      }

      const data = await response.json() as any;
      const choice = data.choices?.[0];

      if (!choice) {
        throw new Error('No response from Together AI');
      }

      return {
        content: choice.message?.content || '',
        model: data.model,
        usage: data.usage
          ? {
              prompt_tokens: data.usage.prompt_tokens,
              completion_tokens: data.usage.completion_tokens,
              total_tokens: data.usage.total_tokens,
            }
          : undefined,
        finish_reason: choice.finish_reason || undefined,
        tool_calls: choice.message?.tool_calls?.map((tc: any) => ({
          id: tc.id,
          type: 'function' as const,
          function: {
            name: tc.function.name,
            arguments: tc.function.arguments,
          },
        })),
        raw: data,
      };
    } catch (error) {
      throw this.formatError(error);
    }
  }

  calculateCost(promptTokens: number, completionTokens: number, model: string): number {
    const modelLower = model.toLowerCase();

    // Find matching model pricing
    let inputRate = 0.2; // Default fallback
    let outputRate = 0.2;

    for (const [modelPrefix, [inp, out]] of Object.entries(TOGETHER_PRICING)) {
      if (modelLower.includes(modelPrefix)) {
        inputRate = inp;
        outputRate = out;
        break;
      }
    }

    // Calculate cost with split pricing
    const inputCost = (promptTokens / 1_000_000) * inputRate;
    const outputCost = (completionTokens / 1_000_000) * outputRate;

    return inputCost + outputCost;
  }

  /**
   * Convert generic messages to Together AI/OpenAI format
   */
  private convertToTogetherMessages(messages: Message[], systemPrompt?: string): any[] {
    const togetherMessages: any[] = [];

    if (systemPrompt) {
      togetherMessages.push({ role: 'system', content: systemPrompt });
    }

    for (const msg of messages) {
      if (msg.role === 'system') {
        togetherMessages.push({ role: 'system', content: msg.content });
      } else if (msg.role === 'user') {
        togetherMessages.push({ role: 'user', content: msg.content });
      } else if (msg.role === 'assistant') {
        togetherMessages.push({ role: 'assistant', content: msg.content });
      } else if (msg.role === 'tool') {
        togetherMessages.push({
          role: 'tool',
          content: msg.content,
          tool_call_id: msg.tool_call_id || '',
        });
      }
    }

    return togetherMessages;
  }

  /**
   * Convert generic tools to OpenAI format (Together AI uses same format)
   */
  private convertTools(tools: Tool[]): any[] {
    return tools.map((tool) => ({
      type: 'function',
      function: {
        name: tool.function.name,
        description: tool.function.description,
        parameters: tool.function.parameters,
      },
    }));
  }
}
